hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferService.java

```diff
@@ -19,43 +19,62 @@
 import com.hazelcast.config.Config;
 import com.hazelcast.config.RingbufferConfig;
 import com.hazelcast.core.DistributedObject;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.partition.strategy.StringPartitioningStrategy;
 import com.hazelcast.ringbuffer.impl.operations.ReplicationOperation;
+import com.hazelcast.spi.DistributedObjectNamespace;
+import com.hazelcast.spi.FragmentedMigrationAwareService;
 import com.hazelcast.spi.ManagedService;
-import com.hazelcast.spi.MigrationAwareService;
 import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.ObjectNamespace;
 import com.hazelcast.spi.Operation;
 import com.hazelcast.spi.PartitionMigrationEvent;
 import com.hazelcast.spi.PartitionReplicationEvent;
 import com.hazelcast.spi.RemoteService;
+import com.hazelcast.spi.ServiceNamespace;
 import com.hazelcast.spi.impl.NodeEngineImpl;
-import com.hazelcast.spi.partition.IPartitionService;
 import com.hazelcast.spi.serialization.SerializationService;
 
+import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Properties;
+import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 
-import static com.hazelcast.partition.strategy.StringPartitioningStrategy.getPartitionKey;
 import static com.hazelcast.spi.partition.MigrationEndpoint.DESTINATION;
 import static com.hazelcast.spi.partition.MigrationEndpoint.SOURCE;
 import static com.hazelcast.util.Preconditions.checkNotNull;
 
 /**
  * The SPI Service that deals with the {@link com.hazelcast.ringbuffer.Ringbuffer}.
  */
-public class RingbufferService implements ManagedService, RemoteService, MigrationAwareService {
-
+public class RingbufferService implements ManagedService, RemoteService, FragmentedMigrationAwareService {
     /**
      * Prefix of ringbuffers that are created for topics. Using a prefix prevents users accidentally retrieving the ringbuffer.
      */
     public static final String TOPIC_RB_PREFIX = "_hz_rb_";
 
+    /**
+     * The ringbuffer service name which defines it in the node engine.
+     */
     public static final String SERVICE_NAME = "hz:impl:ringbufferService";
-    private final ConcurrentMap<String, RingbufferContainer> containers
-            = new ConcurrentHashMap<String, RingbufferContainer>();
+
+    /**
+     * Map from namespace to actual ringbuffer containers. The namespace defines the service and object name which
+     * is the owner of the ringbuffer container.
+     */
+    private final ConcurrentMap<Integer, ConcurrentMap<ObjectNamespace, RingbufferContainer>> containers
+            = new ConcurrentHashMap<Integer, ConcurrentMap<ObjectNamespace, RingbufferContainer>>();
```

Comment:
Does inner map have to be a `ConcurrentMap`?  I guess it's the per partition map.